add blog post about v1 recipe format (#2450)
conda-forge/conda-forge.github.io: news/2025-02-13-conda-forge-v1-recipe-support.md

Reviewer: This should go in `blog/` more then `news/`.

# Announcing the new recipe format on conda-forge

The conda-forge team is excited to announce that the v1 recipe format is available on conda-forge. The v1 recipe format is a community initiative dating back over 3 years to improve the recipe format for conda packages. If you are a maintainer of a feedstock on conda-forge, you have probably dealt with `meta.yaml` files that conda-build utilizes. The file format has some limitations which is why the community has come together to come up with an improved version of the format: the v1 format.

> 

Reviewer: ```suggestion

<!--truncate-->

```

For the "Read more" splits.

## rattler-build and the v1 spec

The v1 recipe format has a number of benefits:

> - It always parses as valid YAML which makes it much easier to modify it with the bot infrastructure of conda-forge. While meta.yaml looks like YAML, it can contain Jinja logic incompatible with the YAML specification, which make significantly complicates parsing and automated manipulation. Additionally, the recipe has a published JSON schema which means that the editing experience in VSCode is greatly improved with contextual help.

Reviewer: ```suggestion
- It always parses as valid YAML which makes it much easier to modify it with the bot infrastructure of conda-forge. While meta.yaml looks like YAML, it can contain Jinja logic incompatible with the YAML specification, which make significantly complicates parsing and automated manipulation. Additionally, the new recipe format has [a published JSON schema](https://github.com/prefix-dev/recipe-format/blob/main/schema.json) which means that the editing experience in VS Code is greatly improved with contextual help.
```

Reviewer 2: "which make significantly" -> "which sifnificantly"

- The new recipe executes much faster, which is partially due to some design decisions (no more recursive evaluation) as well as due to rattler-build building on top of a modern re-implementation of the conda standards in Rust.
- The log output of rattler-build is greatly improved, always showing the user what the final files in the package are and the final metadata.

> - Some features of conda-build, such as multiple outputs, had a lot of implicit behavior. We are fixing that in the v1 recipe.

Reviewer 2: These three bullet points need to separate out two distinct things, namely the format versus the tools that use the format.

Author: Not sure ... this blog post is about the integration in conda-forge and that relies on rattler-build a lot. This is not only about the v0 vs v1 format ...

> You can read much more about the v1 recipe format in the [https://rattler.build](https://rattler.build) docs and [CEPs](https://github.com/conda/ceps/).

Reviewer 2: Let's link explicitly to the two CEPS.

Author: Done

A simple v1 recipe looks something like the following:

```yaml
context:
  # we define named variables in the context instead of `{$ set ... %}` directives
  version: "23.0.0"
 
package:
  name: "boltons"
  # note that we use "GitHub" inspired syntax to access context / Jinja variables
  version: ${{ version }}
 
source:
  url: https://github.com/mahmoud/boltons/archive/refs/tags/${{ version }}.tar.gz
  sha256: 9b2998cd9525ed472079c7dd90fbd216a887202e8729d5969d4f33878f0ff668
 
build:
  noarch: python
  script:
    - python -m pip install . --no-deps -vv
 
requirements:
  host:
    - python
    - pip
    - setuptools
  run:
    - pip
 
about:
  license: BSD-3-Clause
  license_file: LICENSE
```

## How to use the v1 recipe format on conda-forge

If you are adding a new recipe on staged-recipes, then it's easy: just submit a `recipe.yaml` file instead of a `meta.yaml` file.

> In case you already maintain a feedstock, then the conversion can be semi-automated with a tool created by [Hadrien Mary](https://github.com/hadim) called ["feedrattler"](https://github.com/hadim/feedrattler). The tool will take care of the basic conversions steps and uses the `conda-recipe-manager` by Anaconda / Schuyler Martin under the hood to parse the recipe and convert it to the new format.

Reviewer: ```suggestion
In case you already maintain a feedstock, then the conversion can be semi-automated with a tool created by [Hadrien Mary](https://github.com/hadim) called [`feedrattler`](https://github.com/hadim/feedrattler). The tool will take care of the basic conversions steps and uses [`conda-recipe-manager`](https://github.com/conda-incubator/conda-recipe-manager) by Anaconda / [Schuyler Martin](https://github.com/schuylermartin45) under the hood to parse the recipe and convert it to the new format.
```

```shell

> pixi exec feedrattler my-awesome-package-feedstock gh_user

Reviewer: ```suggestion
# One liner with Pixi
pixi exec feedrattler my-awesome-package-feedstock gh_user
# With conda or mamba (as $CONDA)
$CONDA create -n feedrattler feedrattler
$CONDA activate feedrattler
feedrattler my-awesome-package-feedstock gh_user
```

```

To do the conversion by hand, you need to do the following things:

> - In the `conda-forge.yml`, add `conda_build_tool: rattler-build`

Reviewer: ```suggestion
- In the feedstock's `conda-forge.yml`, add `conda_build_tool: rattler-build`
```

- Remove the `meta.yaml` file and add a `recipe.yaml` file following the v1 spec

> - Rerender the feedstock using `conda smithy rerender`

Reviewer: ```suggestion
- Rerender the feedstock using `conda-smithy rerender`
```

- Push your changes to your fork and open a PR to see the CI build your package

## Transition plans

> For the foreseeable future, conda-forge is going to support both formats, v1 and v0. We envision a gradual transition over at least one year to the new spec. There are a couple of places where the v1 spec also needs to finalize / stabilize. One notable place is the `cache` output where CEP discussions are ongoing. This stabilization in the format should happen within Q1 of 2025.

Reviewer: ```suggestion
For the foreseeable future, conda-forge is going to support both formats, v1 and v0. We envision a gradual transition over at least one year to the new spec. There are a couple of places where the v1 spec also needs to finalize / stabilize. One notable place is the `cache`-based multi-output feature, where CEP discussions are ongoing. This stabilization in the format should happen within Q1 of 2025.
```
